### ISSUE TITLE
Match around/inside string pairs

### ISSUE THREAD

**Issue by crabdancing** [helix-editor/helix#8325]
This would allow the user to hit `m a s <string>` to match around that string. Some use cases:

In Nix, multi-line strings are often represented with this sytax:

```nix
rawConfig = ''
  Some code
  some other code
  oh, also here's a single quote in my code -> '
  oooh more code, wow!
'';
```

By being able to hit `m a s ''`, you could quickly match around the entire block reliably. The current mechanism you'd have to use (`m a m` or `m a '` usually erroneously matches only a subsection of the block, as the block itself can contain `'`. Some other cases where quickly matching the space between strings would be useful:

```rs
/* some stuff we don't wanna select */
let a = "some values woo";
// some stuff in the middle we wanna select too
// it's only one line, but let's say it's like 10
let b = "some other values!!";
/* some stuff we don't wanna select */
```

If you selected _between_ the two lets, you could do: `m a s l e t` to quickly match the middle block.

A weak-spot in character matching is the inability to match alphabetical characters. For instance, if you have a string like:

```
abbbbbbbbbbbbbbbbbbafdoudfouh
```

And want to match around or inside the `ab...ba` part -- you cannot currently do so, without writing up a regex, or sighing and resorting to visual mode. With matching around / inside string, you could quickly do:

`m a s a`

And select that entire segment.

There are also other use cases, like weird syntaxes that use names or alphabetical chars to toggle scope.

**Comment by Dav-v** [comment]
Hi, disclaimer: I've used Helix only for a few days.
The first problem that you mentioned could be solved by using tree-sitter and the command "expand_selection" (Expand selection to parent syntax node, `Alt-o` or `Alt-up`). If you are inside the Nix config string and you press `Alt-o` you should select the inner string (the current syntax node in the AST), if you press `Alt-o` again you should expand the selection to around the string. This has the advantage of being syntax-independent so you can use the same keybind in different programming languages without having to rely on the particular syntax structure of the one you're currently using. 
For the second and the third problems I would use multiple selections, for example:

- Select the "area of interest", for example the whole document with `%`
- Use `s` to create multiple selections on each "let"
- Use `Alt-_` to join the above selections into one

This would have the effect of "match around string let". 
For "match inside string let" you could do:

- Select the "area of interest", maybe it could be easier to start from the first "let"
- Use `S` to split the selection on "let"
- Use ( or ) to rotate the primary selection until the one between the two "let" is the primary selection 
- Use `,` to keep only that selection

**Comment by pascalkuthe** [comment]
What exactly would this allow that `mam` doesn't solve? You explanation is hard to follow

**Comment by kirawi** [comment]
From what I understand, this issue is about allowing arbitrary characters to work as pairs. For example `mik` in `k__k` would select the same as `mi"` in `"__"`

**Comment by pascalkuthe** [comment]
> From what I understand, this issue is about allowing arbitrary characters to work as pairs. For example `mik` in `k__k` would select the same as `mi"` in `"__"`

But that is already the case AFAIK? The only place that has hardcoded pairs in `mim` and `mam` (which should work in this case since they will match full TS nodes) but since that list includes `'` that should still work

**Comment by alxpettit** [comment]
> > From what I understand, this issue is about allowing arbitrary characters to work as pairs. For example `mik` in `k__k` would select the same as `mi"` in `"__"`
> 
> But that is already the case AFAIK? The only place that has hardcoded pairs in `mim` and `mam` (which should work in this case since they will match full TS nodes) but since that list includes `'` that should still work

The goal here is to match `''`, not `'`.

**Comment by alxpettit** [comment]
> From what I understand, this issue is about allowing arbitrary characters to work as pairs. For example `mik` in `k__k` would select the same as `mi"` in `"__"`

With my proposal, it would actually be `m i s k`. `mik` would be inadvisable, because future match types may be added which make use of alphabet characters.

**Comment by alxpettit** [comment]
@pascalkuthe 

To understand and replicate the limitation I'm talking about, try the following:

Paste the following into `test.nix`:
```nix
{
  rawConfig = ''
    Some code
    some other code
    oh, also here's a single quote in my code -> '
    oooh more code, wow!
  '';
}
```

Try `m a m`, with the cursor in the interior of the code block. This is what happens:

![image](https://github.com/helix-editor/helix/assets/28266167/e320e643-04b2-41f2-9897-c254044c0067)

This is decidedly _not_ the actual closest surrounding pair. Because `''`..`''` is not considered a 'pair'.

Now, try `m a '`. Again, this is inadequate:

![image](https://github.com/helix-editor/helix/assets/28266167/5c7892c6-3ed2-41d5-9eb8-b1763b35ac51)

**Comment by pascalkuthe** [comment]
Its not that `'''` is not considered a pair but likely that the grammar is structure a bit weird.

I don't think we should introduce an extra text object for string literals for every language. That's a lot of effort uo maintain while being a bit nieche.

We should just improve `mam` to handle this case if possible. 

What version of helix are you using? This may already work on master

**Comment by alxpettit** [comment]
@pascalkuthe 

> I don't think we should introduce an extra text object for string literals for every language. That's a lot of effort uo maintain while being a bit nieche.

Nor do I think we should, either. But `m a s`/`m i s` would be fully general for _all_ weird cases like this.

`m a m` gives insufficient control, and trying to shoehorn it into this use case seems difficult at best! E.g., how does `m a m` know whether to match the inner `'` or the outer `''` if the language only considers the `''` to be relevant? At best, your proposed solution would only work reliably for _known_ languages. But what about unknown or novel languages? I think `m a s` is a _universal_ way of dealing with this use case, in all of them,  while also being self-explanatory and highly discoverable!

May I ask why you're

**Comment by alxpettit** [comment]
I'm picking through the code, and I _think_ @pascalkuthe primary actual concern is implementation complexity?

Which, if so, is fair enough. The current implementation of `select_textobject` and associated wrapper functions, are built with the assumption of a _single_ selection cycle -- when, instead, this allows the user to continuously adjust the bounds of the selection continuously as they type the string, much as can today with the `select` command.

Thus, unlike other `match around/inside` commands, this would require significant adjustment to the flow of logic for `match around/inside`. I guess that leaves the question of whether this design is flexible enough to justify upending the logic of a few functions.

I personally believe that it is, as it allows two matchpoints to be _concisely specified_ in a way that is highly intuitive, covering not only present, but also _future_ use cases. I also don't think that, once implemented, it'll really have much more maintenance overhead than the existing codebase, but I could def be wrong about that? Am curious as to your thoughts on that matter, @pascalkuthe.

I guess if no one wants to work on this, I would be willing to make a few PRs, to see if I can come up with _something_ for this that is satisfactory to everyone. But, if preserving the current `match around / inside` call chain is of significant concern, perhaps a different `select` command that is shorthand for `select around/inside <thing>` could be implemented? Although, IMO, that would be worse UX design -- I would be curious to hear what others think. It could at least be a proof of concept implementation, I suppose.

**Comment by pascalkuthe** [comment]
matching around arbitrary strings requires a textobject query for every single language. W currently support 180 languages. Addying a query for all of them is a lot of effort and needs a lot of maintenance so it should only be done if the benefit is large.

As I said, this should be covered by `mam` and I consider it a bug that it currently doesn't work there. The solution is not to add new textobject but to improve mam

**Comment by alxpettit** [comment]
I do not understand why you think it requires a textobject query for _every single language_.

**Comment by crabdancing** [comment]
I don't know if I'd call this feature request completed, because in my experience it doesn't work as I described. I cannot match around/inside `''`. If I press `ma''` it does not work as it should -- it only pays attention to the first `'`, and then proceeds to wrap around the closest matching pair of single `'`.

Furthermore, it doesn't work in cases where tree-sitter isn't present. We really should have a solution to matching around arbitrary groups/sequences of characters, but if there is one, I'm not aware of it (besides writing some rather obnoxious regex by hand).

It's not as pressing as before for me personally (on account of `nil`, the de facto standard Nix LSP, getting better), but new formats come out every day, and a solution that doesn't require language support for those specific formats would certainly have the most flexibility. For example, just the other week I was mucking about in a [modelfile](https://github.com/ollama/ollama/blob/main/docs/modelfile.md), and found no ability to easily match between two `"""`.

IMO, the solutions currently implemented are not sufficiently general-purpose for a painless workflow.

**Comment by aschoettler** [comment]
I'm a bit of a newbie but I think I figured out how to match inside of a given string. It should work assuming the cursor is between two copies of the search string, which we find via search.

- search backwards for the string `?<string><enter>`
- select both matches `vnv`
- merge selections `<A-minus>` // stop here if you want match-around
- split at string `S<string><enter>`
- discard first match `<A-,>`
- optionally trim whitespace: `_`
